packages/lit-dev-content/samples/examples/react-refs/flying-triangles.ts

```diff
@@ -0,0 +1,261 @@
+import { css, html, LitElement } from 'lit';
+import { customElement, query, state, property } from 'lit/decorators.js';
+
+/*
+  This file is for demo purposes only.
+
+  This module provides a web component that is stateful
+  and uncontrolled similar to video and audio elements.
+
+  The <flying-triangles> web component exposes the following API:
+    methods:
+      - play()
+      - pause()
+
+    attributes:
+      - fps (frames per second)
+    
+    properties:
+      - isPlaying
+
+    events:
+      - 'state-change' 
+*/
+
+const styles = css`
+  :host {
+    max-width: 300px;
+    position: relative;
+    display: block;
+    cursor: pointer;
+  }
+
+  canvas {
+    border: 5px solid #343434;
+  }
+    
+  p {
+    top: 50%;
+    left: 50%;
+    transform: translate(-50%, -50%);
+    position: absolute;
+    opacity: 0.4;
+    margin: 0;
+  }
+`;
+
+@customElement('flying-triangles')
+export class FlyingTriangles extends LitElement {
+  static styles = styles;
+
+  @property({ type: Number }) fps = 24;
+  @state() isPlaying = false;
+  @query('canvas') private canvas!: HTMLCanvasElement;
+
+  private scene = createScene(this.fps);
+  
+  render() {
+    return html`
+      <canvas height="300" width="300"></canvas>
+      <p>${this.isPlaying ? 'click to pause' : 'click to play'}</p>
+    `;
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.addEventListener('click', this.onClick);
+  }
+  
+  disconnectedCallback() {
+    super.connectedCallback();
+    this.removeEventListener('click', this.onClick);
+  }
+
+  updated() {
+    this.scene.fpsAsMS = 1000 / this.fps;
+    this.dispatchEvent(new Event('state-change', { composed: true }));
```

Comment:
This is an anti-pattern. Events should only be dispatched in response to user input. Probably can just move to the click handler?

Reply:
That could work :) I'd rather not do an anti-pattern. I mainly need the `<flying-triangles>` to dispatch the event after state changes so react can reconcile state.

Reply:
Updated component to not use `updated` method.
Made fps logic less ugly.
Moved the `state-change` event to the `onClick` method.

Reply:
Switched to the `willUpdate` method to broadcast a `state-change` event.

I agree there should be some user input associated with events. I think this counts maybe? Setting `isPlaying` or `fps` will update the component state. And those are driven by user input.

The goal here was to satisfy @justinfagnani 's ask for a component to broadcast state changes from a component that could manage its own state.

In this case it's clicking the canvas to pause but i'd imagine a real world use case would be a video that dispatches an event when the video ends. Or a "tick" event every time a timer ticks.

Reply:
where ever is the best place to put this though :)